stone/target/js_types.py

```diff
@@ -0,0 +1,211 @@
+
+        return extra_args_for_types
+
+    def _generate_type(self, data_type, extra_parameters = []):
+        if is_struct_type(data_type):
+            self._generate_struct(data_type, extra_parameters)
+        elif is_union_type(data_type):
+            self._generate_union(data_type)
+
+    def _emit_jsdoc_header(self, doc = None):
+        self.emit()
+        self.emit('/**')
+        if doc:
+            self.emit_wrapped_text(self.process_doc(doc, self._docf), prefix = ' * ')
+
+    def _generate_struct(self, struct_type, extra_parameters = [], nameOverride = None):
+        """
+        Emits a JSDoc @typedef for a struct.
+        """
+        self._emit_jsdoc_header(struct_type.doc)
+        self.emit(' * @typedef {Object} %s' % (nameOverride if nameOverride else fmt_type(struct_type)))
+
+        for param_name, param_type, param_docstring in extra_parameters:
+            param_docstring = ' - %s' % param_docstring if param_docstring else ''
+            self.emit_wrapped_text('@property {%s} %s%s' %
+                (param_type, param_name, param_docstring), prefix = ' * ')
+
+        # NOTE: JSDoc @typedef does not support inheritance. Using @class would be inappropriate,
+        # since these are not nominal types backed by a constructor. Thus, we emit all_fields,
+        # which includes fields on parent types.
+        for field in struct_type.all_fields:
+            field_doc = ' - ' + field.doc if field.doc else ''
+            field_type, nullable, _ = unwrap(field.data_type)
+            field_js_type = fmt_type(field_type)
+            # Translate nullable types into optional properties.
+            field_name = '[' + field.name + ']' if nullable else field.name
+            self.emit_wrapped_text('@property {%s} %s%s' %
+                (field_js_type, field_name, self.process_doc(field_doc, self._docf)),
+                    prefix = ' * ')
+
+        self.emit(' */')
+
+    def _generate_union(self, union_type):
+        """
+        Emits a JSDoc @typedef for a union type.
+        """
+        union_name = fmt_type(union_type)
+        self._emit_jsdoc_header(union_type.doc)
+        self.emit(' * @typedef {Object} %s' % union_name)
+        variant_types = []
+        for variant in union_type.all_fields:
+            variant_types.append("'%s'" % variant.name)
+            variant_data_type, _, _ = unwrap(variant.data_type)
+            # Don't emit fields for void types.
+            if is_void_type(variant_data_type) == False:
+                variant_doc = ' - Available if .tag is %s.' % variant.name
+                if variant.doc:
+                    variant_doc += ' ' + variant.doc
+                self.emit_wrapped_text('@property {%s} [%s]%s' % (fmt_type(variant_data_type), variant.name, variant_doc), prefix = ' * ')
+        jsdoc_tag_union = '(' + '|'.join(variant_types) + ')' if len(variant_types) > 1 else variant_types[0]
```

Comment:
Yup, but TypeScript doesn't actually take JSDoc type annotations into consideration. Previously, I standardized the primitive types to lowercase because it was consistent with both JSDoc documentation and TypeScript. I experimented with a tool that converts JSDoc into a TypeScript definition file (which specifies TypeScript typings for non-TypeScript libraries), but the output was not up to par with what could be achieved with a stone TypeScript definitions generator.

TypeScript will actually support tagged unions in its upcoming 2.0 release (there's a release candidate out now). With this addition, all of Stone's data types map cleanly onto TypeScript's type system! I [wrote a proper TypeScript definition file generator for stone](https://github.com/jvilk/stone/blob/tsd-types-generator/stone/target/tsd_types.py) that generates these tagged unions, but they currently [expose a small bug in TypeScript that will be fixed prior to a proper 2.0 release](https://github.com/Microsoft/TypeScript/issues/10530). Once working, TypeScript will 100% properly tag the arms of a `switch` statement over the `.tag` field, and can determine statically when the switch does not cover all enum variants.

I am happy to open up a PR with the TypeScript typings generator, btw. [I documented what it will cost from a maintenance standpoint to ship TypeScript typings with the Dropbox JS SDK in this issue](https://github.com/dropbox/dropbox-sdk-js/issues/65). If there's no interest from Dropbox, I'll submit the generated types to DefinitelyTyped once the TypeScript bug is fixed, so the people will get the goods either way. :)